### PR TITLE
Make Path3DCollection store indexed offset, and only apply z-ordered offset during draw

### PR DESCRIPTION
**Pull request by hukaidong**
## PR Summary

This PR is targeted to resolve issue [#23155](https://github.com/matplotlib/matplotlib/issues/23155) by maintain offset order for indexing purpose. And z-ordered offset will only be applied during rendering peroid.

## PR Checklist

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by hukaidong** [comment]
@tacaswell would you mind checking this?

File: lib/mpl_toolkits/mplot3d/art3d.py
Comment (tacaswell) [inline lib/mpl_toolkits/mplot3d/art3d.py:752]: Why use `set_offsets` here, but directly setting `self._offsets` below?

File: lib/mpl_toolkits/mplot3d/art3d.py
Comment (tacaswell) [inline lib/mpl_toolkits/mplot3d/art3d.py:745]: Can you make sure that this gets created in `set_3d_properties` as well so that we can be sure to never end up in a situation where the context manager will fail with `AttributeError`?

**Comment by tacaswell** [comment]
This looks reasonable, but had a few comments.  Good use of a context manager.  

Could you please also add a test as well?

**Comment by tacaswell** [comment]
Could you compare this approach to https://github.com/matplotlib/matplotlib/pull/18932 and why you think this is a better path @hukaidong ?

File: lib/mpl_toolkits/mplot3d/art3d.py
Comment (hukaidong) [reply lib/mpl_toolkits/mplot3d/art3d.py:752]: > Why use `set_offsets` here, but directly setting `self._offsets` below?

Here I see set_offsets does a sanity check before apply new offset to the _offset attribute. Since self._offset_zordered is fresh generated, I have it go through the set_offsets methods. On the other hand old_offset is an attribute directly backup from self._offsets, so direct assignment might be a better choice.

File: lib/mpl_toolkits/mplot3d/art3d.py
Comment (hukaidong) [reply lib/mpl_toolkits/mplot3d/art3d.py:745]: > Can you make sure that this gets created in `set_3d_properties` as well so that we can be sure to never end up in a situation where the context manager will fail with `AttributeError`?

I'd rather check the attribute existance by `hasattr` or an `None` check with an initialization with `None` initialize self._offset_zordered with None in initialization to secure the `AttributeError` problem. But, I don't think `set_3d_properties` ever set offset projection. Do you mean we want to execute `do_3d_projection` in the end of `set_3d_properties`?

**Comment by hukaidong** [comment]
> This looks reasonable, but had a few comments. Good use of a context manager.
> 
> Could you please also add a test as well?

Thanks, this does make sense. I'll append a test soon.

**Comment by hukaidong** [comment]
> Could you compare this approach to #18932 and why you think this is a better path @hukaidong ?

I see quite amount of refraction tries to reduce the amount of method provided in artists class, which is not relevant to this topic. So I put them apart. For the work refractoring do_3d_projection, it saves space from storing offset with a cost re-computing projection transform twice. One in `do_3d_projection` in order to compute the over-all z order in the `Axes#draw` method. The other in `Collection#Draw` to get the projection. So, I'd like to say storing an extra offset and swap during drawing time will be cheaper.

File: lib/mpl_toolkits/mplot3d/art3d.py
Comment (tacaswell) [reply lib/mpl_toolkits/mplot3d/art3d.py:745]: Checking for `None` is better than `hasattr` (just because we _can_ dynamically add attributes does not mean we should).

**Comment by hukaidong** [comment]
@tacaswell I'm not sure if this is the best way to test the issue but I've added a test case there. Also a null guard is set in context manager so we won't have to worry about `AttributeError`.

**Comment by hukaidong** [comment]
> > This looks reasonable, but had a few comments. Good use of a context manager.
> > Could you please also add a test as well?
> 
> Thanks, this does make sense. I'll append a test soon.

And also, in case, I've checked main branch and new test case failed there as excepted.

**Review by tacaswell — APPROVED** [review]

**Comment by hukaidong** [comment]
@tacaswell Looks I'm not able to do the merge. Who should I call for the merging?

**Review by ksunden — APPROVED** [review]

**Comment by ksunden** [comment]
@hukaidong just FYI, we do require two approving reviews from core contributors before merging (for changes to library code) which is why @tacaswell did not previously merge this when he first approved it

**Comment by hukaidong** [comment]
> @hukaidong just FYI, we do require two approving reviews from core contributors before merging (for changes to library code) which is why @tacaswell did not previously merge this when he first approved it

Awesome! Thanks for letting me know.

**Pull request merged**
